display-dependency-updates error with 'Wrong segment index: -1' when updated from 2.12.0 to 2.13.0

The error stack trace:
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  30.774 s
[INFO] Finished at: 2022-10-24T10:29:19+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:versions-maven-plugin:2.13.0:display-dependency-updates (default) on project pom: Execution default of goal org.codehaus.mojo:versions-maven-plugin:2.13.0:display-dependency-updates failed: Wrong segment index: -1 -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.mojo:versions-maven-plugin:2.13.0:display-dependency-updates (default) on project pom: Execution default of goal org.codehaus.mojo:versions-maven-plugin:2.13.0:display-dependency-updates failed: Wrong segment index: -1
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:375)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default of goal org.codehaus.mojo:versions-maven-plugin:2.13.0:display-dependency-updates failed: Wrong segment index: -1
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:370)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.IllegalArgumentException: Wrong segment index: -1
    at org.codehaus.mojo.versions.api.Segment.of (Segment.java:45)
    at org.codehaus.mojo.versions.DisplayDependencyUpdatesMojo.lambda$calculateUpdateScope$2 (DisplayDependencyUpdatesMojo.java:700)
    at java.util.Optional.map (Optional.java:260)
    at org.codehaus.mojo.versions.DisplayDependencyUpdatesMojo.calculateUpdateScope (DisplayDependencyUpdatesMojo.java:700)
    at org.codehaus.mojo.versions.DisplayDependencyUpdatesMojo.logUpdates (DisplayDependencyUpdatesMojo.java:716)
    at org.codehaus.mojo.versions.DisplayDependencyUpdatesMojo.execute (DisplayDependencyUpdatesMojo.java:585)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:370)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[ERROR] 
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
```

--- a/src/main/java/org/codehaus/mojo/versions/DisplayDependencyUpdatesMojo.java
+++ b/src/main/java/org/codehaus/mojo/versions/DisplayDependencyUpdatesMojo.java
@@ -226,7 +226,7 @@ public class DisplayDependencyUpdatesMojo
      * @since 2.5
      */
     @Parameter( property = "allowMajorUpdates", defaultValue = "true" )
-    private boolean allowMajorUpdates;
+    private boolean allowMajorUpdates = true;
 
     /**
      * <p>Whether to allow the minor version number to be changed.</p>
@@ -237,7 +237,7 @@ public class DisplayDependencyUpdatesMojo
      * @since 2.5
      */
     @Parameter( property = "allowMinorUpdates", defaultValue = "true" )
-    private boolean allowMinorUpdates;
+    private boolean allowMinorUpdates = true;
 
     /**
      * <p>Whether to allow the incremental version number to be changed.</p>
@@ -249,7 +249,7 @@ public class DisplayDependencyUpdatesMojo
      * @since 2.5
      */
     @Parameter( property = "allowIncrementalUpdates", defaultValue = "true" )
-    private boolean allowIncrementalUpdates;
+    private boolean allowIncrementalUpdates = true;
 
     /**
      * Whether to allow any version change to be allowed. This keeps
@@ -263,7 +263,7 @@ public class DisplayDependencyUpdatesMojo
      */
     @Deprecated
     @Parameter( property = "allowAnyUpdates", defaultValue = "true" )
-    private boolean allowAnyUpdates;
+    private boolean allowAnyUpdates = true;
 
     /**
      * Whether to show additional information such as dependencies that do not need updating. Defaults to false.
@@ -697,7 +697,8 @@ public class DisplayDependencyUpdatesMojo
         return allowAnyUpdates
         ? empty()
         : of( SegmentUtils.determineUnchangedSegment( allowMajorUpdates, allowMinorUpdates,
-                allowIncrementalUpdates, getLog() ).map( s -> Segment.of( s.value() - 1 ) )
+                allowIncrementalUpdates, getLog() )
+                .map( s -> Segment.of( s.value() + 1 ) )
                 .orElse( MAJOR ) );
     }
 

--- a/src/test/java/org/codehaus/mojo/versions/DisplayDependencyUpdatesMojoTest.java
+++ b/src/test/java/org/codehaus/mojo/versions/DisplayDependencyUpdatesMojoTest.java
@@ -165,7 +165,8 @@ public class DisplayDependencyUpdatesMojoTest extends AbstractMojoTestCase
                     new StubArtifactResolver( new ArtifactStubFactory(), false, false ) )
             {{
                 setProject( createProject() );
-                setVariableValueToObject( this, "allowMinorUpdates", true );
+                setVariableValueToObject( this, "allowAnyUpdates", false );
+                setVariableValueToObject( this, "allowMajorUpdates", false );
                 setVariableValueToObject( this, "processDependencies", true );
                 setVariableValueToObject( this, "dependencyIncludes",
                         singletonList( WildcardMatcher.WILDCARD ) );
@@ -179,6 +180,132 @@ public class DisplayDependencyUpdatesMojoTest extends AbstractMojoTestCase
                     not( anyOf( containsString( "2.0.0-SNAPSHOT" ),
                             containsString( "2.0.0-beta" ),
                             containsString( "2.0.0-rc1" ) ) ) );
+        }
+        finally
+        {
+            if ( tempPath != null && Files.exists( tempPath ) )
+            {
+                Files.delete( tempPath );
+            }
+        }
+    }
+
+    @Test
+    public void testAllowMajorUpdatesFalse()
+            throws MojoExecutionException, MojoFailureException, IllegalAccessException, IOException
+    {
+        Path tempPath = null;
+        try
+        {
+            tempPath = Files.createTempFile( "display-dependency-updates", "" );
+            final File tempFile = tempPath.toFile();
+            new DisplayDependencyUpdatesMojo( mockRepositorySystem(),
+                    null, mockArtifactMetadataSource( new HashMap<String, String[]>()
+            {{
+                put( "default-dependency", new String[] {"1.0.0", "1.1.0", "2.0.0"} );
+            }} ), null,
+                    new StubArtifactResolver( new ArtifactStubFactory(), false, false ) )
+            {{
+                setProject( createProject() );
+                setVariableValueToObject( this, "allowAnyUpdates", false );
+                setVariableValueToObject( this, "allowMajorUpdates", false );
+                setVariableValueToObject( this, "processDependencies", true );
+                setVariableValueToObject( this, "dependencyIncludes",
+                        singletonList( WildcardMatcher.WILDCARD ) );
+                setVariableValueToObject( this, "dependencyExcludes", emptyList() );
+                this.outputFile = tempFile;
+                setPluginContext( new HashMap<>() );
+            }}.execute();
+
+            String output = String.join( "", Files.readAllLines( tempPath ) );
+
+            assertThat( output, containsString( "1.1.0" ) );
+            assertThat( output, not( containsString( "2.0.0" ) ) );
+        }
+        finally
+        {
+            if ( tempPath != null && Files.exists( tempPath ) )
+            {
+                Files.delete( tempPath );
+            }
+        }
+    }
+
+    @Test
+    public void testAllowMinorUpdatesFalse()
+            throws MojoExecutionException, MojoFailureException, IllegalAccessException, IOException
+    {
+        Path tempPath = null;
+        try
+        {
+            tempPath = Files.createTempFile( "display-dependency-updates", "" );
+            final File tempFile = tempPath.toFile();
+            new DisplayDependencyUpdatesMojo( mockRepositorySystem(),
+                    null, mockArtifactMetadataSource( new HashMap<String, String[]>()
+            {{
+                put( "default-dependency", new String[] {"1.0.0", "1.0.1", "1.1.0", "2.0.0"} );
+            }} ), null,
+                    new StubArtifactResolver( new ArtifactStubFactory(), false, false ) )
+            {{
+                setProject( createProject() );
+                setVariableValueToObject( this, "allowAnyUpdates", false );
+                setVariableValueToObject( this, "allowMinorUpdates", false );
+                setVariableValueToObject( this, "processDependencies", true );
+                setVariableValueToObject( this, "dependencyIncludes",
+                        singletonList( WildcardMatcher.WILDCARD ) );
+                setVariableValueToObject( this, "dependencyExcludes", emptyList() );
+                this.outputFile = tempFile;
+                setPluginContext( new HashMap<>() );
+            }}.execute();
+
+            String output = String.join( "", Files.readAllLines( tempPath ) );
+
+            assertThat( output, containsString( "1.0.1" ) );
+            assertThat( output, not( containsString( "1.1.0" ) ) );
+            assertThat( output, not( containsString( "2.0.0" ) ) );
+        }
+        finally
+        {
+            if ( tempPath != null && Files.exists( tempPath ) )
+            {
+                Files.delete( tempPath );
+            }
+        }
+    }
+
+    @Test
+    public void testAllowIncrementalUpdatesFalse()
+            throws MojoExecutionException, MojoFailureException, IllegalAccessException, IOException
+    {
+        Path tempPath = null;
+        try
+        {
+            tempPath = Files.createTempFile( "display-dependency-updates", "" );
+            final File tempFile = tempPath.toFile();
+            new DisplayDependencyUpdatesMojo( mockRepositorySystem(),
+                    null, mockArtifactMetadataSource( new HashMap<String, String[]>()
+            {{
+                put( "default-dependency", new String[] {"1.0.0", "1.0.0-1", "1.0.1", "1.1.0", "2.0.0"} );
+            }} ), null,
+                    new StubArtifactResolver( new ArtifactStubFactory(), false, false ) )
+            {{
+                setProject( createProject() );
+                setVariableValueToObject( this, "allowAnyUpdates", false );
+                setVariableValueToObject( this, "allowIncrementalUpdates", false );
+                setVariableValueToObject( this, "processDependencies", true );
+                setVariableValueToObject( this, "dependencyIncludes",
+                        singletonList( WildcardMatcher.WILDCARD ) );
+                setVariableValueToObject( this, "dependencyExcludes", emptyList() );
+                this.outputFile = tempFile;
+                setPluginContext( new HashMap<>() );
+            }}.execute();
+
+            String output = String.join( "", Files.readAllLines( tempPath ) );
+
+            assertThat( output, containsString( "1.0.0-1" ) );
+            assertThat( output, not( containsString( "1.0.1" ) ) );
+            assertThat( output, not( containsString( "1.1.0" ) ) );
+            assertThat( output, not( containsString( "2.0.0" ) ) );
         }
         finally
         {
@@ -207,7 +334,8 @@ public class DisplayDependencyUpdatesMojoTest extends AbstractMojoTestCase
                     new StubArtifactResolver( new ArtifactStubFactory(), false, false ) )
             {{
                 setProject( createProject() );
-                setVariableValueToObject( this, "allowIncrementalUpdates", true );
+                setVariableValueToObject( this, "allowAnyUpdates", false );
+                setVariableValueToObject( this, "allowMinorUpdates", false );
                 setVariableValueToObject( this, "processDependencies", true );
                 setVariableValueToObject( this, "dependencyIncludes",
                         singletonList( WildcardMatcher.WILDCARD ) );
@@ -228,6 +356,47 @@ public class DisplayDependencyUpdatesMojoTest extends AbstractMojoTestCase
             {
                 Files.delete( tempPath );
             }
+        }
+    }
+
+    @Test
+    public void testDetermineUpdatedSegment() throws Exception
+    {
+        File outputFile = null;
+        try
+        {
+            outputFile = File.createTempFile( "display-dependency-updates", "" );
+            assert outputFile.exists();
+
+            DisplayDependencyUpdatesMojo mojo = (DisplayDependencyUpdatesMojo) mojoRule.lookupConfiguredMojo(
+                    new File( "target/test-classes/org/codehaus/mojo/display-dependency-updates/ruleset" ),
+                    "display-dependency-updates" );
+
+            assertThat( mojo.ruleSet, notNullValue() );
+            assertThat( mojo.ruleSet.getIgnoreVersions(), notNullValue() );
+            assertThat( mojo.ruleSet.getIgnoreVersions(), Matchers.hasSize( 3  ) );
+            assertThat( mojo.ruleSet.getIgnoreVersions(), hasItem( matches(
+                    new TestIgnoreVersions().withVersion( "1.0.1" ) ) ) );
+            assertThat( mojo.ruleSet.getIgnoreVersions(), containsInAnyOrder(
+                    matches( new TestIgnoreVersions().withVersion( "1.0.1" ) ),
+                    matches( new TestIgnoreVersions().withType( TYPE_REGEX ).withVersion( ".+-SNAPSHOT" ) ),
+                    matches( new TestIgnoreVersions().withType( TYPE_REGEX ).withVersion( ".+-M\\d+" ) ) ) );
+
+            // This is just an example of how to create it-style tests as unit tests; the advantage is easier debugging
+            mojo.outputFile = outputFile;
+            mojo.artifactMetadataSource = mockArtifactMetadataSource( new HashMap<String, String[]>()
+            {{
+                put( "dummy-api", new String[] { "1.0.0", "1.0.1", "1.1.0-M1", "1.2.0-SNAPSHOT" } );
+            }} );
+
+            assertThat( mojo.ruleSet.getIgnoreVersions(), Matchers.hasSize( 3 ) );
+            mojo.execute();
+            List<String> output = Files.readAllLines( outputFile.toPath(), UTF_8 );
+            assertThat( output, not( hasItem( containsString( "1.1.0-M1" ) ) ) );
+        }
+        finally
+        {
+            assert outputFile == null || !outputFile.exists() || outputFile.delete();
         }
     }
 }

